libs/esp32boot/esp32init.erl

```diff
@@ -0,0 +1,79 @@
+%
+% This file is part of AtomVM.
+%
+% Copyright 2023 Davide Bettio <davide@uninstall.it>
+%
+% Licensed under the Apache License, Version 2.0 (the "License");
+% you may not use this file except in compliance with the License.
+% You may obtain a copy of the License at
+%
+%    http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS,
+% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+% See the License for the specific language governing permissions and
+% limitations under the License.
+%
+% SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+%
+
+-module(esp32init).
+
+-export([start/0]).
+
+start() ->
+    console:print(<<"AtomVM init.\n">>),
+    boot().
+
+maybe_start_dev_mode(SystemStatus) ->
+    case {SystemStatus, esp:nvs_get_binary(atomvm, dev_mode)} of
+        {_, <<"always">>} ->
+            ep32devmode:start_dev_mode();
+        {_, <<"never">>} ->
+            not_started;
+        {ok, undefined} ->
+            not_started;
+        {failed_app_start, undefined} ->
+            esp32devmode:start_dev_mode()
+    end.
+
+% TODO: add support for multiple apps
+% /dev/partition/by-name/app1.avm
+% /dev/partition/by-name/app2.avm
+
+boot() ->
+    BootPath = get_boot_path(),
+    case atomvm:add_avm_pack_file(BootPath, [{name, app}]) of
+        ok ->
+            StartModule = get_start_module(),
+            maybe_start_dev_mode(ok),
+            io:format("Starting application...~n"),
+            StartModule:start();
+        {error, Reason} ->
+            io:format("Failed app start: ~p.~n", [Reason]),
+            maybe_start_dev_mode(failed_app_start)
+    end.
```

Comment:
There is a bit of a logical error in what is going on here.

You removed the code that sleeps forever if dev mode is enabled.  That is a problem in the base case where you have erased flash and are starting the device, because dev mode is started, but the VM then immediately terminates:
```
---
AtomVM init.
I (895) sys: Loaded BEAM partition main.avm at address 0x210000 (size=1048576 bytes)
Failed app start: invalid_avm.
AtomVM finished with return value: started
I (985) AtomVM: AtomVM application terminated.  Going to sleep forever ...
```

You will need to tell the VM to sleep after dev mode is started (`timer:sleep(infinity)`), but you also want to support the case of running in dev mode even if the user has an application (presumably, if the user has set `atomvm:devmode` to `always` in NVS.  So you are going to need to signal to the `boot` function when to sleep forever, while also NOT sleeping forever if the user is not using demoed.

Reply:
I changed the flow:
- when devmode is set to "never", app start function is tail-called, so the behavior is nearly identical to the previous
- otherwise try ... catch block is used, and devmode is started according to settings.

Reply:
👍🏽